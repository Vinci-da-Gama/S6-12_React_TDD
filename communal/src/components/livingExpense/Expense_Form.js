import React, { Component } from 'react';
import moment from 'moment';
import { SingleDatePicker } from 'react-dates';
import { Button, Alert } from 'reactstrap';

class ExpenseForm extends Component {
    constructor(props) {
        super(props);
        this.state = {
            description: props.expense ? props.expense.description : '',
            note: props.expense ? props.expense.note : '',
            amount: props.expense ? (props.expense.amount / 100).toString() : '',
            createdAt: props.expense ? moment(props.expense.createdAt) : moment(),
            calendarFocused: false,
            error: ''
        };
    }

    onDescriptionChange(e) {
        const description = e.target.value;
        this.setState(() => ({ description }));
    }

    onAmountChange(e) {
        const amount = e.target.value;
        if (!amount || amount.match(/^\d{1,}(\.\d{0,2})?$/)) {
            this.setState(() => ({ amount }));
        }
    }

    onDateChange(createdAt) {
        if (createdAt) {
            this.setState(() => ({ createdAt }));
        }
    }

    onFocusChange({ focused }) {
        this.setState(() => ({ calendarFocused: focused }));
    }

    onNoteChange(e) {
        const note = e.target.value;
        this.setState(() => ({ note }));
    }

    onFormSubmit(e) {
        e.preventDefault();
        const error = 'Please provide description and amount.';
        if (!this.state.description || !this.state.amount) {
            this.setState(() => ({ error }));
        } else {
            this.setState(() => ({ error: '' }));
            this.props.onSubmit({
                description: this.state.description,
                amount: parseFloat(this.state.amount, 10) * 100,
                createdAt: this.state.createdAt.valueOf(),
                note: this.state.note
            });
        }
    }
    
    render () {
        return (
            <div>
                { this.state.error && <Alert color="danger">{this.state.error}</Alert> }
                <form noValidate name="expenseForm" onSubmit={ (event) => { this.onFormSubmit(event) } }>
                    <input type="text" name="ef_description" placeholder="Description..." autoFocus
                        value={this.state.description} onChange={ (event) => { this.onDescriptionChange(event) }} />
                    <input type="text" name="ef_Amount" placeholder="Amount..."
                        value={this.state.amount} onChange={ (event) => { this.onAmountChange(event) } } />
                    <SingleDatePicker date={this.state.createdAt} focused={this.state.calendarFocused}
                        onDateChange={ this.onDateChange } onFocusChange={ this.onFocusChange }
                        numberOfMonths={1} isOutsideRange={ () => false } />
                    <textarea name="ef_note" rows="5"
                        placeholder="Add a note for your expense (optional)"
                        value={this.state.note}
                        onChange={ (event) => this.onNoteChange(event) }></textarea>
                    <Button color="success" outline block>
                        Add_Expense
                    </Button>
                </form>
            </div>
        );
    }
};

export default ExpenseForm;