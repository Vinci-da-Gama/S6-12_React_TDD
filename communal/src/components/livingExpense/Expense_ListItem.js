import React from 'react';
import { Link } from 'react-router-dom';
import moment from 'moment';
import numeral from 'numeral';
import { ListGroup, ListGroupItem, Badge, Label } from 'reactstrap';

const ExpenseListItem = ({ id, description, amount, createdAt }) => {
    return (
        <div>
            <ListGroup>
                <ListGroupItem className="justify-content-between">
                    <Link to={`/edit/${id}`}>
                        <h3>{description}</h3>
                    </Link>
                    <i>Date: </i>{moment(createdAt).format('MMMM Do, YYYY')}
                    <Badge pill>
                        <Label>Amount: </Label>
                        {numeral(amount / 100).format('$0,0.00')}
                    </Badge>
                </ListGroupItem>
            </ListGroup>
        </div>
    );
};

export default ExpenseListItem;
