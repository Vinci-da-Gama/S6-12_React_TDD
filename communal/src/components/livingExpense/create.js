import React, { Component } from 'react';

class CreateExpense extends Component {
    render () {
        return (
            <div>
                create exp....
            </div>
        );
    }
};

export default CreateExpense;