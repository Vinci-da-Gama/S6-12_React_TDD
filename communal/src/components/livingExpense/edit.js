import React, { Component } from 'react';

class EditExpense extends Component {
    render () {
        return (
            <div>
                edit exp....
            </div>
        );
    }
};

export default EditExpense;