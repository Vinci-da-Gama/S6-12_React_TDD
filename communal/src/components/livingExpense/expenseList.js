import React, { Component } from 'react';
import { connect } from 'react-redux';
import { Button, Alert } from 'reactstrap';

import selectExpenses from '../../helpers/data-handler/expense-filterNsort';
import ExpenseListItem from './Expense_ListItem';

export class ExpenseListCompo extends Component {
    render() {
        if (this.props.expenses.length === 0 || !this.props.expenses) {
            return (
                <Alert color="danger">
                    No expense item list.
                </Alert>
            );
        }
        return (
            <div>
                {
                    this.props.expenses.map((expense, idx) => {
                        return <ExpenseListItem key={ expense.id+idx } {...expense} />
                    })
                }
            </div>
        );
    }
}

const mapStateToProps = (state) => ({
    expenses: selectExpenses(state.expenses, state.setValViaFilters)
});

/* const mapDispatchToProps = { 
} */

export default connect(mapStateToProps, null)(ExpenseListCompo);
