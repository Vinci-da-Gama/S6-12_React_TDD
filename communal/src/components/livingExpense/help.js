import React, { Component } from 'react';

class HelpExpense extends Component {
    render () {
        return (
            <div>
                help exp....
            </div>
        );
    }
};

export default HelpExpense;