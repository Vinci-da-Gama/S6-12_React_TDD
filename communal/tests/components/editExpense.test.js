import React from 'react';
import { shallow } from 'enzyme';

import fixtureExpenses from '../fixtures/expenses.fixture';
import { EditExpenseCompo } from '../../src/components/livingExpense/editExpense';

let editSpy, removeSpy, history, match, wrapper, e2;

beforeEach(() => {
    editSpy = jest.fn();
    removeSpy = jest.fn();
    history = {
        push: jest.fn()
    };
    // due to we set fixed id for fixtureExpense, we actually donot touch this match.
    // thus use jest.fn();
    e2 = fixtureExpenses[2];
    match = {
        url: `/edit/${e2.id}`
    };
    wrapper = shallow(<EditExpenseCompo editExpense={editSpy} removeExpense={removeSpy} history={history} match={match} expense={e2} />);
});

describe('Test Edit Expense.', () => {

    test('should render edit expense page correctly.', () => {
        expect(wrapper).toMatchSnapshot();
    });
    
    test('should handle edit expense.', () => {
        wrapper.find('ExpenseForm').prop('onSubmit')(e2);
        expect(editSpy).toHaveBeenLastCalledWith(e2.id, e2);
        expect(history.push).toHaveBeenLastCalledWith('/');
    });
    
    test('should handle remove expense.', () => {
        wrapper.find('Button').simulate('click');
        expect(removeSpy).toHaveBeenLastCalledWith({id: e2.id});
        expect(history.push).toHaveBeenLastCalledWith('/');
    });

});