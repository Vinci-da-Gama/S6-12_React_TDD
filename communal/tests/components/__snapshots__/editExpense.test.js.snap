// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Test Edit Expense. should render edit expense page correctly. 1`] = `
<Row
  tag="div"
>
  <Card
    body={true}
    inverse={true}
    tag="div"
  >
    <CardHeader
      className="text-secondary"
      tag="h2"
    >
      Edit_Expense
    </CardHeader>
    <CardBody
      tag="div"
    >
      <ExpenseForm
        currentUrl="/edit/3"
        expense={
          Object {
            "amount": 4500,
            "createdAt": 345600000,
            "description": "Credit_Card",
            "id": "3",
            "note": "Pay back credit",
          }
        }
        onSubmit={[Function]}
      />
      <Col
        className="offset-0 offset-sm-3 offset-md-4 my-3"
        md="4"
        sm="6"
        tag="div"
        widths={
          Array [
            "xs",
            "sm",
            "md",
            "lg",
            "xl",
          ]
        }
        xs="12"
      >
        <Button
          block={true}
          className="rounded-0"
          color="danger"
          onClick={[Function]}
          outline={true}
          tag="button"
        >
          Remove_Expense
        </Button>
      </Col>
    </CardBody>
    <CardFooter
      className="text-muted"
      tag="div"
    >
      Footer
    </CardFooter>
  </Card>
</Row>
`;
